db/comdb2.h

```diff
@@ -1098,14 +1098,16 @@ struct client_query_path_component {
     int nfind;
     int nnext;
     int nwrite;
+    int nblob;
+    int unused; // 4-byte padding for cache-line alignment
```

Comment:
What field are you trying to align to the cache-line size? Also it's only useful when multiple cores are accessing the same data structure, which I do not believe is the case here.

Reply:
There is a compile time assert which fails because the struct is padded to align with 8bytes. Hence the extra 4 byte field

Reply:
First of all, this is just an alignment which has nothing to do with the cache line. Second of all, the struct itself seems perfectly aligned without explicitly padding with an int. We should find out where exactly the compile time assertion fails, and pad that struct instead.

Reply:
IIRC structs are padded to prevent the linker from putting anything else in the last cache line of the structure. This ensures that the entire cache line is for the structure itself, and it is not subject to being invalidated by access to unrelated data. 
Check out the BB_COMPILE_TIME_ASSERT below. That fails because the sizeof() returns the size AFTER compiler padding. Sure I can change the Length of the structure to match the padded size without actually adding an extra field to the struct , but I believe this is more readable as it is now explicit that an extra field has been added to account for padding.

Reply:
You do not need to pad `struct client_query_path_component`. Like I said, all members of the struct are already aligned to their own size. The assertion below hence would not fail even without the extra field.

> IIRC structs are padded to prevent the linker from putting anything else in the last cache line of the structure.

I'm not sure about this, either. A struct is padded to a multiply of the size of its widest member. That's all.

I believe this is ultimately what you need to make this sane. Note that `client_query_stats` needs to account for padding in the end, because `client_query_stats` needs to be padded to the size of `double cost` (once again, its widest member), which is 8 bytes.

```diff
diff --git a/db/comdb2.h b/db/comdb2.h
index 06ced4dd..c7c26a56 100644
--- a/db/comdb2.h
+++ b/db/comdb2.h
@@ -1014,6 +1014,7 @@ struct client_query_path_component {
     int nfind;
     int nnext;
     int nwrite;
+    int nblobs;
     char table[32];
 #if 0
     [32+1/*.*/+31/*dbname*/];
@@ -1021,7 +1022,7 @@ struct client_query_path_component {
     int ix;
 };

-enum { CLIENT_QUERY_PATH_COMPONENT_LEN = 4 + 4 + 4 + (1 * 32) + 4 };
+enum { CLIENT_QUERY_PATH_COMPONENT_LEN = 4 + 4 + 4 + 4 + (1 * 32) + 4 };

 BB_COMPILE_TIME_ASSERT(client_query_path_component_len,
                        sizeof(struct client_query_path_component) ==
@@ -1074,12 +1075,13 @@ struct client_query_stats {
     int n_components;
     double cost;
     struct client_query_path_component path_stats[1];
+    int padding;
 };

 enum {
     CLIENT_QUERY_STATS_PATH_OFFSET = 4 + 4 + 4 + 4 + (4 * 16) + 4 + 4 + 8,
     CLIENT_QUERY_STATS_LEN =
-        CLIENT_QUERY_STATS_PATH_OFFSET + CLIENT_QUERY_PATH_COMPONENT_LEN
+        CLIENT_QUERY_STATS_PATH_OFFSET + CLIENT_QUERY_PATH_COMPONENT_LEN + 4
 };
```

Reply:
Understood! 
So, since client_query_stats uses client_query_path_component , the responsibility for alignment should fall on client_query_stats. makes sense. Will incorporate the changes!

Reply:
We apparently do not have an assertion for sizeof client_query_stats. You might want to add one.

Reply:
> So, since client_query_stats uses client_query_path_component , the responsibility for alignment should fall on client_query_stats.

No quite. the compiler will align a struct regardless. It's just in this case, no padding is needed for `client_query_path_component` (since all members are aligned to their own size), while padding is needed for `client_query_stats` (because the size of the struct needs to be a multiple of 8 bytes).

Reply:
I think we're saying the same thing. You can pad client_query_path_component (which would increase it's size to 56, thereby keeping the size of client_query_stats a multiple of 8) , or pad client_query_stats to ensure that it's a multiple of 8. It's a better choice to pad the latter as you suggested, as that is exactly what the compiler does anyways.